File: lib/wikipediafs/fs.rb
# WikipediaFS
# Copyright (C) 2005 - 2006  Mathieu Blondel
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA

require 'gettext'

module WikipediaFS

module PathUtils

    def split_path(path)
        dirname = File.dirname(path)
        basename = File.basename(path)
        if dirname == basename
            ["/", "."]
        else
            [dirname, basename]
        end
    end

    def scan_path(path)
        path.scan(/[^\/]+/)
    end

end

SwapFile = Struct.new(:str)

class Directory
    include PathUtils

    attr_accessor :name
    attr_reader :parent

    def initialize(site=nil)
        @site = site
        @files = {"." => self}
        @swap_files = {}
    end

    def parent=(par)
        @parent = par
        @files[".."] = par
    end

    def dir_directory?(filename)
        not @files[filename].nil? and @files[filename].kind_of? Directory
    end

    def [](filename)
        @files[filename]
    end

    def dir_files
        names = @files.keys.sort
        names.find_all do |n| 
            n != "." and n != ".." and not @files[n].kind_of? SwapFile
        end
    end

    def dir_file?(filename)
        if @files[filename].nil?
            if filename =~ /\.mw$/
                # If file is a .mw file and but does not exist,
                # we get the page and check if it has content or not
                page_name = filename[0..-4]
                page = Page.new(@site, page_name)

                if page.str.empty?
                    false
                else
                    @files[filename] = page
                    true
                end
            else
                false
            end
        else
            not @files[filename].kind_of?(Directory)
        end
    end

    def dir_read_file(filename)
        @files[filename].str
    end

    def dir_size(filename)
        @files[filename].str.length
    end

    def dir_can_write?(filename)
        if /\.mw$/ =~ filename or /\~$/ =~ filename or \
           /\.swp$/ =~ filename or /^\./ =~ filename
            true
        else
            false
        end
    end

    def dir_write_to(filename, str)
        unless @files.include? filename
            if /\.mw$/ =~ filename
                page_name = filename[0..-4]
                @files[filename] = Page.new(@site, page_name)
            else
                @files[filename] = SwapFile.new
            end
        end

        unless str.strip.empty?
            @files[filename].str = str
        end
    end

    def dir_can_mkdir?(filename)
        false
    end

    def dir_mkdir(dirname)
    end

    def dir_mkdir_obj(dirname, dir_obj)
        @files[dirname] = dir_obj
        @files[dirname].parent = self
    end

    def dir_can_delete?(filename)
        true
    end

    def dir_delete(filename)
        @files.delete(filename) if @files.include? filename
    end

end

class Root < Directory
    include GetText

    def initialize
        super
        parent = self
        @conf = Config.instance
        @log = Logger.instance

        @conf.sites.each do |dirname, site|
            dir_mkdir_obj(dirname, Directory.new(site))
        end
    end
  
    def get_file(path)
        dirs = scan_path(path)

        last = self
        dirs.each do |dir|
            last = last[dir]
            return nil if last == nil
        end
        last
    end
    private :get_file

    def get_parent_directory(path)
        dirname, basename = split_path(path)
        get_file(dirname)
    end
    private :get_parent_directory

    def directory?(path)
        dirname, basename = split_path(path)
        dir = get_parent_directory(path)
        ret = dir.dir_directory?(basename)
        @log.info(_("Checking if %s is a directory (%s)") % [path, ret])
        ret
    end

    def contents(path)
        @log.info(_("Listing files at %s") % path)
        get_file(path).dir_files
    end

    def file?(path)
        dirname, basename = split_path(path)
        dir = get_parent_directory(path)
        ret = dir.dir_file?(basename)
        @log.info(_("Checking if %s is a file (%s)") % [path, ret])
        ret
    end

    def read_file(path)
        @log.info(_("Reading file %s") % path)
        dirname, basename = split_path(path)
        get_parent_directory(path).dir_read_file(basename)
    end

    def size(path)
        dirname, basename = split_path(path)
        dir = get_parent_directory(path)      
        ret = dir.dir_size(basename)
        @log.info(_("Getting size of %s (%d)") % [path, ret])
        ret
    end

    def can_write?(path)
        dirname, basename = split_path(path)
        dir = get_parent_directory(path)      
        ret = dir.dir_can_write?(basename)
        @log.info(_("Checking if can write %s (%s)") % [path, ret])
        ret
    end

    def write_to(path, str)
        @log.info(_("Writing to %s") % path)
        dirname, basename = split_path(path)
        dir = get_parent_directory(path)
        dir.dir_write_to(basename, str)
    end

    def can_delete?(path)
        dirname, basename = split_path(path)
        dir = get_parent_directory(path)      
        ret = dir.dir_can_delete?(basename)
        @log.info(_("Checking if can delete %s (%s)") % [path, ret])
        ret
    end

    def delete(path)
        @log.info(_("Writing to %s") % path)
        dirname, basename = split_path(path)
        dir = get_parent_directory(path)
        dir.dir_delete(basename)
    end

    def can_mkdir?(path)
        dirname, basename = split_path(path)
        dir = get_parent_directory(path)
        ret = dir.dir_can_mkdir?(basename)
        @log.info(_("Checking if can make dir %s (%s)") % [path, ret])
        ret
    end

    def mkdir(path)
        @log.info(_("Writing to %s") % path)
        dirname, basename = split_path(path)
        dir = get_parent_directory(path)
        dir.dir_mkdir(basename)
    end

    def can_rmdir?(path)
        dirname, basename = split_path(path)
        dir = get_parent_directory(path)
        ret = dir.dir_can_rmdir?(basename)
        @log.info(_("Checking if can remove dir %s (%s)") % [path, ret])
        ret
    end

    def rmdir(path)
        @log.info(_("Removing dir %s") % path)
        dirname, basename = split_path(path)
        dir = get_parent_directory(path)
        dir.dir_rmdir(basename)
    end

end

end

if __FILE__ == $0
    require './config'
    require './logger'
    require './page'
    require 'pp'

    root = WikipediaFS::Root.new
    pp root.contents("/mblondel.org")
    pp root.file?("/mblondel.org/Japon")
    pp root.file?("/mblondel.org/Japon.mw")
    pp root.read_file("/mblondel.org/Japon.mw")
end

File: src/wikipediafs/fs.py
# -*- coding: utf-8 -*-

# WikipediaFS
# Copyright (C) 2005 - 2007 Mathieu Blondel
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import os.path, re, time
from metadir import MetaDir
from config import CONFIG
from article import Article
from user import User
from logger import LOGGER

class ArticleDir:
    def __init__(self, fs, config):
        self.fs = fs
        self.config = config
        self.login_time = 0

        self.files = {}
        self.dirs = {}

    def get_article_full_name(self, path):
        # Returns article name. This can include subpages.
        return '/'.join(path.split("/")[2:])

    def get_article_file_name(self, path):
        # File name as used in the directory i.e. without the subpages.
        return os.path.basename(path)

    def is_valid_file(self, path):
        file_name = self.get_article_file_name(path)

        for forbidden_char in ('#', '<', '>', '[', ']', '|', '{', '}'):
            if file_name.count(forbidden_char) > 0:
                return False
        
        if file_name[-3:] == ".mw":
            return True
        else:
            return False

    def contents(self, path):
        arr = self.dirs.keys()
        for k in self.files.keys():
            if not self.files[k].is_empty:
                arr.append(k)
        return arr
            
    def is_directory(self, path):
        name = self.get_article_file_name(path)
        return self.dirs.has_key(name)

    def is_file(self, path):
        if self.is_valid_file(path):
            txt = self.read_file(path)
            if len(txt.strip()) == 0:
                return False
            else:
                return True
        else:
            return False

    def set_cookie_string(self, force):
        if force or not self.config.has_key("cookie_str"):
            if self.config["username"] is not None and \
               self.config["password"] is not None:
                user = User(logger=LOGGER, **self.config)
                cookie_str = user.getCookieString()
                self.config["cookie_str"] = cookie_str

            self.login_time = time.time()

    def get_art(self, path):
        file_name = self.get_article_file_name(path)
        full_name = self.get_article_full_name(path)

        if int(time.time()) - self.login_time > CONFIG.login_cache_time:
            self.set_cookie_string(1)
        else:
            if self.files.has_key(file_name):
                return self.files[file_name]
            else:
                self.set_cookie_string(0)

        art = Article(full_name[0:-3], # removes .mw from the name
                      cache_time=CONFIG.cache_time,
                      logger = LOGGER,
                      **self.config)
        self.files[file_name] = art

        return art

    def read_file(self, path):
        art = self.get_art(path)
        txt = art.get()
        return txt

    def write_to(self, path, txt):
        art = self.get_art(path);
        return art.set(txt)

    def size(self, path):
        LOGGER.debug("FSdir size %s" % (path))
        return len(self.read_file(path))

    def mtime(self, path):
        art = self.get_art(path)
        # Do a get here just so we have a current Edittime.
        art.get()
        tmp = art.wpEdittime
        t = time.mktime((int(tmp[0:4]), int(tmp[4:6]), int(tmp[6:8]), int(tmp[8:10]), int(tmp[10:12]), int(tmp[12:21]), 0, 0, -1))
        if time.daylight:
            t -= time.altzone
        else:
            t -= time.timezone

        return t
                    
    def mode(self, path):
        LOGGER.debug("FSdir mode %s" % (path))
        return 0755
                 
    def unlink(self, path):
        LOGGER.debug("FSdir unlink %s" % (path))
        file_name = self.get_article_file_name(path)
        if self.files.has_key(file_name):
            self.files.pop(file_name)
            return True # succeeded
        else:
            return False
                            
    def mkdir(self, path):
        LOGGER.debug("FSdir mkdir %s" % (path))
        name = self.get_article_file_name(path)
        self.dirs[name] = True        
        self.fs.set_dir(path, ArticleDir(self.fs, self.config))
        return True                               

class Root:
    def __init__(self, fs):
        self.fs = fs
        self.dirs = {}

        self.regexp = "(wikipedia|wiktionary|wikiquote|wikibooks|wikisource)"
        self.regexp += "\-([a-z]{2})"
        self.regexp = re.compile(self.regexp, re.IGNORECASE)

        for dirname, config in CONFIG.sites.items():
            self.dirs[dirname] = True
            self.fs.set_dir("/" + dirname, ArticleDir(self.fs, config))

    def contents(self, path):
        if path == "/":
            return self.dirs.keys()
        else:
            return []

    def is_directory(self, path):
        basename = os.path.basename(path)
        if path == "/" or self.dirs.has_key(basename):
            return True
        else:
            return False

    def is_file(self, path):
        return False # There is no file at the root

    def mode(self, path):
        return 0755

    def mkdir(self, path):
        # add a site from the wikimedia foundation
        basename = os.path.basename(path)
        match = self.regexp.match(basename)

        # add it only if a directory with the same name does not exist yet
        if match and not self.is_directory(path):
            lang = match.group(2)
            site = match.group(1)
            config = {
                "host": "%s.%s.org" % (lang, site),
                "basename" : "/w/index.php",
                "username" : None,
                "password" : None
            }
            self.dirs[basename] = True
            self.fs.set_dir("/" + basename, ArticleDir(self.fs, config))

            return True
        else:
            return False


class WikipediaFS(MetaDir):
    def __init__(self, *arr, **dic):
        MetaDir.__init__(self, *arr, **dic)
        self.set_root(Root(self))


if __name__ == "__main__":
    server = WikipediaFS(version="%prog VERSION",
                     usage='blabla',
                     dash_s_do='setsingle')

    server.parse(errex=1)
    server.multithreaded = 0
    server.main()
